### PR TITLE
Optimize ORC reading when file contains many tiny stripes

### LINKED ISSUE #3663: Optimize ORC reader when reading small stripes

### PR DESCRIPTION
**Pull request by haozhun**
Fixes #3663

### CODE REVIEW

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        OrcDataSource actual = CachingOrcDataSource.wrapIfTinyStripes(FakeOrcDataSource.INSTANCE,
+                ImmutableList.of(new StripeInformation(123, 3, 10, 10, 10), new StripeInformation(123, 33, 10, 10, 10)));
+        assertNotInstanceOf(actual, CachingOrcDataSource.class);
+
+        actual = CachingOrcDataSource.wrapIfTinyStripes(FakeOrcDataSource.INSTANCE,
+                ImmutableList.of(new StripeInformation(123, 3, 10, 10, 10), new StripeInformation(123, 33, 10, 10, 10), new StripeInformation(123, 63, 10, 10, 10)));
+        assertInstanceOf(actual, CachingOrcDataSource.class);
+
+        actual = CachingOrcDataSource.wrapIfTinyStripes(FakeOrcDataSource.INSTANCE,
+                ImmutableList.of(new StripeInformation(123, 3, 10, 10, 10), new StripeInformation(123, 33, 10, 10, 10), new StripeInformation(123, 63, 1048576 * 8 - 20, 10, 10)));
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Add parens

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    @Test
+    public void testIntegration()
+            throws IOException
+    {
+        DelegateOrcDataSource orcDataSource = new DelegateOrcDataSource(
+                new FileOrcDataSource(tempFile.getFile(), new DataSize(1, Unit.MEGABYTE), new DataSize(1, Unit.MEGABYTE), new DataSize(1, Unit.MEGABYTE)));
+        OrcReader orcReader = new OrcReader(orcDataSource, new OrcMetadataReader());
+        assertEquals(orcDataSource.getCount(), 1); // read file tail
+        List<StripeInformation> stripes = orcReader.getFooter().getStripes();
+        assertGreaterThanOrEqual(stripes.size(), 3); // Sanity check number of stripes. Normally is equal to 3. But can be 8 on Travis.
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: "This can be three or higher, depending on ..."

And list the actual factor (CPU count, available memory, etc.) that causes it to vary rather than a specific test environment.

Why does the stripe count vary? Is this related to the Hive ORC memory manager?

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    public static <T, U extends T> void assertNotInstanceOf(T actual, Class<U> expectedType)
+    {
+        assertNotNull(actual, "actual is null");
+        assertNotNull(expectedType, "expectedType is null");
+        if (expectedType.isInstance(actual)) {
+            fail(String.format("expected:<%s> to not be an instance of <%s>", actual, expectedType.getName()));
+        }
+    }
+
+    static FileSinkOperator.RecordWriter createOrcRecordWriter(File outputFile, Format format, Compression compression, ObjectInspector columnObjectInspector)
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: I think this can be private

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java (hunk 1 above)
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Add parens everywhere there is multiplication and addition/subtraction

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+        @Override
+        public <K> Map<K, FixedLengthSliceInput> readFully(Map<K, DiskRange> diskRanges)
+                throws IOException
+        {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class DelegateOrcDataSource
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: "Delegating" is the preferred form, but let's use "Forwarding" which the term that Guava uses and also used in a couple other places in Presto.

Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Pull this up into a top level class in the testing tree

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                throws IOException
+        {
+            // do nothing
+        }
+
+        @Override
+        public <K> Map<K, FixedLengthSliceInput> readFully(Map<K, DiskRange> diskRanges)
+                throws IOException
+        {
+            throw new UnsupportedOperationException();
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Why does this throw but the others don't? Seems like they should all have the same behavior

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    private static class DelegateOrcDataSource
+            implements OrcDataSource
+    {
+        private int count;
+        private List<DiskRange> lastReadRanges;
+        private OrcDataSource delegate;
+
+        public DelegateOrcDataSource(OrcDataSource delegate)
+        {
+            this.delegate = delegate;
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Check for null

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        }
+
+        public int getCount()
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Name this `getReadCount`

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        tableProperties.setProperty("orc.stripe.size", "1200000");
+
+        return new OrcOutputFormat().getHiveRecordWriter(
+                jobConf,
+                new Path(outputFile.toURI()),
+                Text.class,
+                compression != NONE,
+                tableProperties,
+                () -> { }
+        );
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Put this on the previous line

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    private static final int POSITION_COUNT = 50000;
+
+    private TempFile tempFile;
+
+    @BeforeClass
+    public void setUp()
+            throws Exception
+    {
+        tempFile = new TempFile("presto_test_cods", "orc");
+        Random random = new Random();
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Inline this and use `ThreadLocalRandom`

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        this.stripeIterator = requireNonNull(stripes, "stripes is null").listIterator();
+    }
+
+    public static OrcDataSource wrapIfTinyStripes(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        if (stripes.size() < 3) {
+            return dataSource;
+        }
+        for (StripeInformation stripe : stripes) {
+            if (stripe.getDataLength() + stripe.getFooterLength() + stripe.getIndexLength() > MAX_CACHE_LENGTH) {
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Add parens

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        return dataSource.getSize();
+    }
+
+    @VisibleForTesting
+    void readCacheAt(long offset)
+            throws IOException
+    {
+        // Assumption: stripes are always read in order
+        // Assumption: bytes that are not part of any stripe is never read
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: "are never read"

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java (hunk 10 above)
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Add a helper method `stripeLength(stripe)`

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+            long nextStripeLength = nextStripe.getIndexLength() + nextStripe.getDataLength() + nextStripe.getFooterLength();
+            long newEndOffset = nextStripe.getOffset() + nextStripeLength;
+            if (newEndOffset - startOffset > MAX_CACHE_LENGTH) {
+                stripeIterator.previous();
+                break;
+            }
+            endOffset = newEndOffset;
+        }
+        cacheLength = Ints.checkedCast(endOffset - startOffset);
+        expandCacheIfNeeded(cacheLength);
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: This is only used in one place, and it is directly tied to `cacheLength`, so it would be clear to inline it.

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        readFully(position, buffer, 0, buffer.length);
+    }
+
+    @Override
+    public void readFully(long position, byte[] buffer, int bufferOffset, int length)
+            throws IOException
+    {
+        if (position < cachePosition) {
+            throw new IllegalArgumentException(String.format("read request (offset %d length %d) is before cache (offset %d length %d)", position, length, cachePosition, cacheLength));
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Use `checkArgument`

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        }
+        if (cache == null || position > cachePosition + cacheLength) {
+            readCacheAt(position);
+        }
+        if (position + length > cachePosition + cacheLength) {
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: This check happens after you read. Is this actually verifying an invariant (i.e. it is an assertion)? If so, use `verify()`

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+public class CachingOrcDataSource
+        implements OrcDataSource
+{
+    private static final long MAX_CACHE_LENGTH = new DataSize(8, DataSize.Unit.MEGABYTE).toBytes();
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: This should be a constructor argument.  I would use `maxMergeDistance` in the existing Hive code.

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+    public CachingOrcDataSource(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        this.stripeIterator = requireNonNull(stripes, "stripes is null").listIterator();
+    }
+
+    public static OrcDataSource wrapIfTinyStripes(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        if (stripes.size() < 3) {
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Why did we choose 3 here?  I would assume that we would want to do this for even a single stripe since it will coalesce all reads into a single read.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java (hunk 5 above)
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: It not needed here. Therefore it's not implemented.

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java (hunk 13 above)
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Discussed in person. Agreed that it's fine as is.

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java (hunk 14 above)
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java]: Discussed in person. Agreed that it's fine as is.

**Comment by haozhun** [comment]
@dain Addressed comments.

**Comment by dain** [comment]
I think we should take a slightly different approach to this.  When we are creating the `CachingOrcDataSource` we can determine the cache regions and simply pass them into constructor.  That way the `CachingOrcDataSource` doesn't need to understand stripe details at all which should really simplify the code. 

Additionally, I think we can add another simple optimization.  In the `OrcReader` constructor we can check the overall file size and if it is smaller than `maxMergeDistance`, we can create s single `CachingOrcDataSource` for the whole file.

**Comment by haozhun** [comment]
Decoupled `CachingOrcDataSource` with stripes.

**Comment by haozhun** [comment]
ping @dain

**Comment by haozhun** [comment]
@dain back to you

**Comment by dain** [comment]
Looks good
